### PR TITLE
Add method to parse plain text from request body

### PR DESCRIPTION
**Pull request by samtgarson**
This PR abstracts the logic to parse the raw buffer from a request so it can be used by `json` and also a new `text` method enabling parsing of plain text from request.

### CODE REVIEW

**Comment by timneutkens** [comment]
Duplicate of #116. @leo what's your take on this?

**Comment by samtgarson** [comment]
My bad, didn't see that one. I'm testing micro with a really tiny node server which is rendering JS components server side for a Rails app, so this would be really handy. Would be interested to know  your thoughts on this.

(I'm moving from express with `bodyParser.text()`)

**Comment by timneutkens** [comment]
Personally I don't mind having it in there. Though most people will use either `json` or another format parser, for example: I created an [urlencoded-body-parser](https://github.com/timneutkens/urlencoded-body-parser) for use with micro.

**Comment by samtgarson** [comment]
For sure—I figured it was so similar to `json` it would be a useful helper and easier to include here.

**Comment by timneutkens** [comment]
@leo @rauchg could you take a look at this one?

File: lib/index.js [hunk 1]
```diff
@@ -39,18 +40,14 @@ async function run(req, res, fn, onError) {
...
     const encoding = typer.parse(type).parameters.charset
-    const str = await getRawBody(req, {limit, length, encoding})
+    const buff = await getRawBody(req, {limit, length, encoding})
 
-    try {
-      return JSON.parse(str)
-    } catch (err) {
-      throw createError(400, 'Invalid JSON', err)
-    }
+    return buff
```
Comment (timneutkens) [inline lib/index.js:50]: you  can `return await getRawBody(req, {limit, length, encoding})`

**Comment by rauchg** [comment]
This makes a lot of sense to me

**Comment by rauchg** [comment]
The question is, how common is this, that it merits having it in core?

**Comment by timneutkens** [comment]
Just published https://www.npmjs.com/package/raw-text

**Comment by rauchg** [comment]
I'd use `raw-text` from @timneutkens. Maybe feature it on the README ?

**Comment by timneutkens** [comment]
@rauchg I will add an example to the new Readme 👍

**Comment by samtgarson** [comment]
👌

**Comment by dimapaloskin** [comment]
@timneutkens @rauchg @samtgarson 
`raw-text`  module is ok for most cases but there is one problem.

Look the code:
```js
const { json } = require('micro');
const parse = require('raw-text');

module.exports = async (req, res) => {
  const textBody = await parse(req);
  const jsonBody = await json(req);
  return { hello: 'world' };
};
```
json(req) will never be fulfilled because `raw-text` doesn't cache final buffer in the similar place with `json`.

Look this two lines please:
https://github.com/zeit/micro/blob/master/lib/server.js#L63
https://github.com/timneutkens/raw-text/blob/master/index.js#L7

`json` method keep buffer into map so we can use it as many times as we please.
`raw-text` keep it right in request (req.rawBody) and we can use it several times too.

But someone will have a problems in case if he want to use **both** this methods.

So I see two solutions:
1) Create `text` method and expose it like `json`.
2) Rewrite current current `json` code and keep raw body in the request.

What do you think about that? This is make sense?

**Comment by samtgarson** [comment]
I would suggest:

```js
const parse = require('raw-text');

module.exports = async (req, res) => {
  const textBody = await parse(req);
  const jsonBody = JSON.parse(textBody);
  return { hello: 'world' };
};
```

**Comment by dimapaloskin** [comment]
@samtgarson Yes, I understand how I can get the JSON from raw text. The main problem is caching. Let's think that someone need get raw body twice. Or just someone did grab all stream data before `json` calling. There is no any chance get the body again :(

In my code I need to use getRawBody before as "req" object get inside micro-service handler function. So inside this function I can't use `json` or `raw-text` or something else thats works with request stream. Because stream was read already. 

Will be very useful to make a ability to works with cache (or just raw body directly).

**Comment by samtgarson** [comment]
I'm going to delegate to @timneutkens because I'm no expert (and I also don't understand why you'd need to use both methods in the same request... 😕)

**Comment by dimapaloskin** [comment]
@samtgarson 
Something like that:

```js
const { run }  = require('micro');
const getRawBody = require('raw-body');

const file = 'path/to/other/micro.js';

module.exports = async (req, res) => {
  const body = await getRawBody(req);
  // there is some body transformation stuff
  const mod = require(resolve(file));
  run(req, res, mod);
}
```

**Comment by dimapaloskin** [comment]
@samtgarson 
> I'm going to delegate to @timneutkens because I'm no expert

No problem. Thank you in any case!

**Comment by timneutkens** [comment]
Okay, so the case here is that initially our solution to the multiple calls issues was adding `req.rawBody`. Later rauchg removed this and added the map. I wonder if we could just export the rawBodyMap for external modules to depend upon. That would certainly solve this issue.

**Comment by dimapaloskin** [comment]
@timneutkens I don't think that rawBodyMap exporting is a good solution. At least it is looks a bit strange :) We can implement something like that:
```js
exports.getCachedBody = (req) => {
  return rawBodyMap.get(req);
}
```
But it is looks strange too because `.text` method is more logical approach.

Also I interesting the reason why rauchg change `req.rawBody` to map. Maybe it's can be reverted if it does not carry any critical fixes or improvements.

**Comment by timneutkens** [comment]
> Also I interesting the reason why rauchg change req.rawBody to map. Maybe it's can be reverted if it does not carry any critical fixes or improvements.

this is because we don't want to add stuff to the req variable. Me might have to add the text method.
Lets open this ticket for now. cc @leo @rauchg

**Comment by timneutkens** [comment]
Closing this in favor of a new issues I'm going to create right now.

**Pull request closed without merging**